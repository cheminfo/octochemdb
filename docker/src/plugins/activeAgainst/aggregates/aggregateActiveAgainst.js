import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import debugLibrary from '../../../utils/Debug.js';
/**
 * @description Aggregate the active against data from the activeOrNaturals collection
 * @param {*} connection
 * @output {collection} activeAgainst collection
 */
export async function aggregate(connection) {
  const debug = debugLibrary('aggregateActiveAgainst');
  try {
    // Get collections from the database
    const options = { collection: 'activeAgainst', connection };
    const progress = await connection.getProgress(options.collection);
    const progressOfSourceCollection = await connection.getProgress(
      'activesOrNaturals',
    );
    const collectionActivesOrNaturals = await connection.getCollection(
      'activesOrNaturals',
    );
    const sources = md5(progressOfSourceCollection);
    // Add logs to the collection importLogs
    const logs = await connection.getImportationLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });
    // Get the last document imported
    const lastDocumentImported = await getLastDocumentImported(
      options.connection,
      progress,
      options.collection,
    );

    if (
      lastDocumentImported === null ||
      sources !== progress.sources ||
      progress.state !== 'updated'
    ) {
      // If the last document imported is null, or the sources are different, or the state is not updated, then we need to import the data
      const temporaryCollection = await connection.getCollection(
        `${options.collection}_tmp`,
      );
      debug.info('start Aggregation process');
      progress.state = 'aggregating';
      await connection.setProgress(progress);
      // Aggregate the data from the activesOrNaturals collection
      let dbRefs = await collectionActivesOrNaturals
        .aggregate(
          [
            { $project: { activities: '$data.activities' } },
            { $match: { activities: { $ne: [] } } },
            {
              $project: {
                _id: 0,
                activities: '$activities',
              },
            },
          ],
          {
            allowDiskUse: true, // allow aggregation to use disk if necessary
            maxTimeMS: 60 * 60 * 1000, // 1h
          },
        )
        .toArray();
      let start = Date.now();
      let count = 0;
      let total = dbRefs.length;

      dbRefs = dbRefs.filter((dbRef) => dbRef.activities !== undefined);

      for (let dbRef of dbRefs) {
        for (let activity of dbRef.activities) {
          let collectionActivity = await connection.getCollection(
            activity.collection,
          );
          let cursor = await collectionActivity.find({
            _id: activity.oid,
          });
          if (await cursor.hasNext()) {
            let entryActivity = await cursor.next();
            if (entryActivity.data?.targetTaxonomies) {
              for (let taxonomy of entryActivity.data.targetTaxonomies) {
                let entry = {
                  data: {},
                };
                if (taxonomy.superkingdom) {
                  entry.data.superkingdom = taxonomy.superkingdom;
                }
                if (taxonomy.kingdom) entry.data.kingdom = taxonomy.kingdom;
                if (taxonomy.phylum) entry.data.phylum = taxonomy.phylum;
                if (taxonomy.class) entry.data.class = taxonomy.class;
                if (taxonomy.order) entry.data.order = taxonomy.order;
                if (taxonomy.family) entry.data.family = taxonomy.family;
                if (taxonomy.genus) entry.data.genus = taxonomy.genus;
                let levels = Object.keys(entry.data);
                let id = '';
                for (let level of levels) {
                  id += `${entry.data[level]}`;
                }
                entry._id = id.replace(/\s/g, '');
                let cursor = await temporaryCollection.find({ _id: entry._id });
                if (!(await cursor.hasNext())) {
                  entry.count = 1;
                  await temporaryCollection.updateOne(
                    { _id: entry._id },
                    { $set: entry },
                    { upsert: true },
                  );
                } else {
                  await temporaryCollection.updateOne(
                    { _id: entry._id },
                    { $inc: { count: 1 } },
                    { upsert: true },
                  );
                }
                count++;
              }
            }
          }
        }
        if (Date.now() - start > Number(process.env.DEBUG_THROTTLING)) {
          await debug.info(`Aggregation progress: ${count}/${total}`);
          start = Date.now();
        }
      }

      // rename temporary collection
      await temporaryCollection.rename(options.collection, {
        dropTarget: true,
      });
      // update logs
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'aggregated';
      await connection.updateImportationLog(logs);
      // update progress
      progress.sources = sources;
      progress.dateEnd = Date.now();
      progress.state = 'aggregated';
      await connection.setProgress(progress);
      await debug.info('Aggregation Done');
    } else {
      await debug.info(`Aggregation already up to date`);
    }
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'activeAgainst',
        connection,
        stack: e.stack,
      });
    }
  }
}
