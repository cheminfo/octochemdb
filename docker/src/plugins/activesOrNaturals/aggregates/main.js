import { cpus } from 'os';
import { Worker } from 'worker_threads';

import debugLibrary from '../../../utils/Debug.js';

export async function main(links) {
  const debug = debugLibrary('improveActivesOrNaturals Main');

  try {
    const values = Object.values(links).sort(() => Math.random() - 0.5);
    const workers = [];
    const url = new URL('worker.js', import.meta.url);

    const numWorkers = cpus().length / 2;
    const chunkSize = Math.floor(values.length / numWorkers);

    const chunks = [];
    for (let i = 0; i < numWorkers; i++) {
      const start = i * chunkSize;
      const end = i === numWorkers - 1 ? values.length : (i + 1) * chunkSize;
      const chunk = values.slice(start, end);
      chunks.push(chunk);
    }
    for (let i = 0; i < numWorkers; i++) {
      let worker = new Worker(url);
      let links = chunks[i];
      let workerID = `worker ${i + 1}`;
      worker.postMessage({ links, workerID });
      workers.push(worker);
    }

    await Promise.all(
      workers.map(
        (worker) =>
          new Promise((resolve, reject) => {
            worker.on('message', (message) => {
              debug(message);
              resolve(message);
            });
            worker.on('error', reject);
            worker.on('exit', (code) => {
              if (code !== 0) {
                reject(new Error(`Worker stopped with exit code ${code}`));
              }
            });
          }),
      ),
    );
  } catch (e) {
    debug(e);
  }
}
