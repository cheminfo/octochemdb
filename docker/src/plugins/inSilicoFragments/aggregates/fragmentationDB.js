export const fragmentationDB = {
  ionization: {
    positive: [
      {
        Label: 'Ionization',
        rxnCode: 'eF`BLGtX!eF``fLGt#Qd Qd#!R@FL]Rgp !R@FL]vgp',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eM`BN`~b@!eM``fN`~c@#Q[ Q[#!R@AM?DquRo@ !R@AM?Dqtzo@',
      },

      {
        Label: 'Ionization',
        rxnCode: 'gCh@AGj@`!gChA@Icu@_PP#qTq qTq#!RbOvw?_x@GYK| !RbOvw?_x@GWk|',
      },
      {
        Label: 'Ionization',
        rxnCode:
          'gCa@@dmPFtwhHcMCAlp!eMH`eIhOhp#Qv@ Qv#!R@AL@[@@SGSH| !R@Fp@Dqwz``',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eFB@HcA}D@!eFB`HIcA}F@#QX QX#!R_vp]bgp !R_vp]vgp',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eMB@HchH!eMB`HIchOh`#Q[ Q[#!R@AM?Dqtbo@ !R@AM?Dquz@`',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eMHAIXMicLLG~r!eFH`fJGtP#QX QX#!R@AL@[AtbO@ !R@AL]nkp',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eMHBN``!eMH`fN`~b@#Qg Qg#!R@FL?XqtJ_@ !R@FL?XqwZ_@',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eFB@HcA}D@!eFB`HIcA}F@#QX QX#!R_vp]bgp !R_vp]vgp',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eFHBLGtP!eFH`fLGtX#QX QX#!R@AL]Pmp !R@AL]^gp',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eM`BN`~b@!eM``fN`~bOza@#Qg Qg#!R@Fq?[AuJ?@ !R@FM?Dqvz_@',
      },
    ],
  },
  positive: {},
};
