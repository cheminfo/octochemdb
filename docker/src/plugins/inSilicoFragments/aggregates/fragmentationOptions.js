export let fragmentationOptions = {
  maxDepth: 3,
  limitReactions: 500,
  minIonizations: 1,
  maxIonizations: 1,
  minReactions: 0,
  maxReactions: 2,
};
