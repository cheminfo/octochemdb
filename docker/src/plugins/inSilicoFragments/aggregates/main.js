import { cpus } from 'os';
import { Worker } from 'worker_threads';

import debugLibrary from '../../../utils/Debug.js';
import { OctoChemConnection } from '../../../utils/OctoChemConnection.js';

const connection = new OctoChemConnection();
export async function main(links) {
  const debug = debugLibrary('improveInSilicoFragments Main');

  try {
    let total = Object.keys(links).length;

    let values = Object.values(links).sort(() => Math.random() - 0.5);

    const workers = [];
    const url = new URL('worker.js', import.meta.url);

    const numWorkers = cpus().length / 2;
    const chunkSize = Math.floor(values.length / numWorkers);
    debug.trace(`Starting ${numWorkers} workers`);
    debug.trace(`Chunk size: ${chunkSize}`);
    debug.trace(`Total: ${total}`);
    for (let i = 0; i < numWorkers; i++) {
      const start = i * chunkSize;
      const end = i === numWorkers - 1 ? values.length : (i + 1) * chunkSize;
      const chunk = values.slice(start, end);
      let worker = new Worker(url);
      worker.postMessage({ links: chunk, workerID: i });
      workers.push(worker);
    }

    const counts = new Uint32Array(numWorkers);
    let lastLogDate = Date.now();
    await Promise.all(
      workers.map(
        (worker) =>
          new Promise((resolve, reject) => {
            worker.on('message', (message) => {
              counts[message.workerID] = message.currentCount;
              if (
                Date.now() - lastLogDate >
                  Number(process.env.DEBUG_THROTTLING) &&
                message.status === 'running'
              ) {
                let current = counts.reduce(
                  (previous, current) => previous + current,
                  0,
                );
                lastLogDate = Date.now();
                debug.trace(`Fragmented: ${current} / ${total} Molecules`);
              }
              if (message.status === 'done') {
                resolve(message);
              }
            });
            worker.on('error', (err) => {
              debug.fatal(err.message, {
                collection: 'inSilicoFragments',
                connection,
                stack: err.stack,
              });
            });
            worker.on('exit', (code) => {
              debug.trace(code);
              if (code !== 0) {
                reject(new Error(`Worker stopped with exit code ${code}`));
              }
            });
          }),
      ),
    );
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'inSilicoFragments',
        connection,
        stack: e.stack,
      });
    }
  }
}
