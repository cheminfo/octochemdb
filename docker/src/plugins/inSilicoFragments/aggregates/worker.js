import { parentPort } from 'worker_threads';

import { reactionFragmentation } from 'mass-fragmentation';
import OCL from 'openchemlib';

import debugLibrary from '../../../utils/Debug.js';
import { OctoChemConnection } from '../../../utils/OctoChemConnection.js';

//import { fragmentationDB } from './fragmentationDB.js';

const { Molecule } = OCL;
const connection = new OctoChemConnection();
const debug = debugLibrary('WorkerProcess');

parentPort?.on('message', async (dataEntry) => {
  let warnCount = 0;
  let warnDate = Date.now();
  try {
    const { links, workerID } = dataEntry;
    debug.trace(`Worker ${workerID} started`);
    // get worker number
    const temporaryCollection = await connection.getCollection(
      `inSilicoFragments_tmp`,
    );
    let count = 0;
    let start = Date.now();

    for (const link of links) {
      try {
        let result = {
          _id: link.id,
          data: {
            ocl: { idCode: link.idCode },
          },
        };

        let molecule = Molecule.fromIDCode(link.idCode);
        if (molecule.getAtoms() <= 200) {
          const fragmentationOptions = {
            ionizationKind: ['esiPositive'],
            maxDepth: 3,
            limitReactions: 500,
            minIonizations: 1,
            maxIonizations: 1,
            minReactions: 0,
            maxReactions: 2,
          };

          // @ts-ignore
          debug.trace(`Fragmenting ${molecule.toSmiles()}`);
          let fragments = reactionFragmentation(molecule, fragmentationOptions);
          const massesArray = getMasses(fragments.masses);
          if (massesArray?.length > 0) {
            result.data.masses = { positive: massesArray };
            await temporaryCollection.updateOne(
              { _id: link.id },
              { $set: result },
              { upsert: true },
            );
          }
        }

        count++;
        if (Date.now() - start > Number(process.env.DEBUG_THROTTLING)) {
          parentPort?.postMessage({
            workerID,
            currentCount: count,
            status: 'running',
          });
          start = Date.now();
        }
      } catch (e) {
        if (connection) {
          warnCount++;
          if (Date.now() - warnDate > Number(process.env.DEBUG_THROTTLING)) {
            await debug.warn(
              `Warning(fragmentation) happened ${warnCount}:${e.message} `,
              {
                collection: 'inSilicoFragments',
                connection,
                stack: e.stack,
              },
            );
          }
          warnDate = Date.now();
        }
      }
    }
    parentPort?.postMessage({ workerID, currentCount: count, status: 'done' });
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'inSilicoFragments',
        connection,
        stack: e.stack,
      });
    }
  }
});

function getMasses(masses) {
  let result = {};
  for (let i = 0; i < masses.length; i++) {
    if (masses[i]?.mz) {
      result[masses[i].mz] = true;
    }
  }
  return Object.keys(result).map(Number);
}
