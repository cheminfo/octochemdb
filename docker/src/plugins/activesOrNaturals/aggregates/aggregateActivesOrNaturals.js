import md5 from 'md5';

import debugLibrary from '../../../utils/Debug.js';
import getCollectionsLinks from '../utils/getCollectionsLinks.js';

import { main } from './main.js';
/**
 * @description Aggregate all synchronized collections into one collection of Bioactivities or/and Natural Products
 * @param {*} connection MongoDB connection
 * @return {Promise} Returns ActiveOrNaturals collection
 */
export async function aggregate(connection) {
  const collectionNames = [
    'lotuses',
    'npasses',
    'npAtlases',
    'cmaups',
    'coconuts',
    'bioassays',
  ];
  const debug = debugLibrary('aggregateActivesOrNaturals');
  try {
    const options = { collection: 'activesOrNaturals', connection };
    // Get progress,logs, target, lastDocument and links of the collection
    const progress = await connection.getProgress(options.collection);
    const targetCollection = await connection.getCollection(options.collection);

    let { links, collectionSources } = await getCollectionsLinks(
      connection,
      collectionNames,
    );
    const pubmedProgress = await connection.getProgress('pubmeds');
    collectionSources.push(pubmedProgress.sources);
    const sources = md5(collectionSources);
    const logs = await connection.getImportationLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });

    if (sources !== progress.sources || progress.state !== 'aggregated') {
      // if lastDocumentImported is null or sources are different from the progress, start aggregation process
      const temporaryCollection = await connection.getCollection(
        `${options.collection}_tmp`,
      );

      // debug unique numbers of noStereoTautomerIDs
      debug.trace(
        `Unique numbers of noStereoTautomerIDs: ${Object.keys(links).length}`,
      );
      debug.info('start Aggregation process');
      // set progress to aggregating

      progress.state = 'aggregating';
      await connection.setProgress(progress);
      // parse all noStereoTautomerIDs and get their info
      await main(links);
      // rename temporary collection to activesOrNaturals
      await temporaryCollection.rename(options.collection, {
        dropTarget: true,
      });
      // set logs to aggregated
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'aggregated';
      await connection.updateImportationLog(logs);
      // set progress to aggregated
      progress.sources = sources;
      progress.dateEnd = Date.now();
      progress.state = 'aggregated';
      await connection.setProgress(progress);

      // Create Indexes
      await targetCollection.createIndex({ 'data.em': 1 });
      await targetCollection.createIndex({ 'data.mf': 1 });
      await targetCollection.createIndex({ 'data.kwBioassays': 1 });
      await targetCollection.createIndex({ 'data.meshTerms': 1 });
      await targetCollection.createIndex({ 'data.kwTaxonomies': 1 });
      await targetCollection.createIndex({ 'data.nbActivities': 1 });
      await targetCollection.createIndex({ 'data.pubmeds': 1 });
      await targetCollection.createIndex({ 'data.nbPubmeds': 1 });
      await targetCollection.createIndex({ 'data.patents': 1 });
      await targetCollection.createIndex({ 'data.nbPatents': 1 });
      await targetCollection.createIndex({ 'data.nbTaxonomies': 1 });
      await targetCollection.createIndex({ 'data.nbMassSpectra': 1 });
      await targetCollection.createIndex({ 'data.molecules': 1 });
      await targetCollection.createIndex({ 'data.nbMolecules': 1 });

      debug.info('Aggregation Done');
    } else {
      debug.info(`Aggregation already up to date`);
    }
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'activesOrNaturals',
        connection,
        stack: e.stack,
      });
    }
  }
}
