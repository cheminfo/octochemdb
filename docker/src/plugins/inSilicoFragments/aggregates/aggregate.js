import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import debugLibrary from '../../../utils/Debug.js';
import getCollectionsLinks from '../utils/getCollectionsLinks.js';

import { main } from './main.js';

export async function aggregate(connection) {
  const debug = debugLibrary('inSilicoFragments');
  try {
    // Get collections from the database
    const options = { collection: 'inSilicoFragments', connection };
    const progress = await connection.getProgress(options.collection);
    const progressOfSourceCollection = await connection.getProgress(
      'activesOrNaturals',
    );

    const sources = md5(progressOfSourceCollection);
    // Add logs to the collection importLogs
    const logs = await connection.getImportationLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });
    // Get the last document imported
    const lastDocumentImported = await getLastDocumentImported(
      options.connection,
      progress,
      options.collection,
    );

    if (
      lastDocumentImported === null ||
      sources !== progress.sources ||
      progress.state !== 'updated'
    ) {
      const temporaryCollection = await connection.getCollection(
        `${options.collection}_tmp`,
      );
      debug.info('start fragmentation process');
      progress.state = 'aggregating';
      await connection.setProgress(progress);

      let links = await getCollectionsLinks(connection);

      await main(links);
      await temporaryCollection.createIndex({ 'data.masses.positive': 1 });
      await temporaryCollection.createIndex({ 'data.ocl.idCode': 1 });
      // rename temporary collection
      await temporaryCollection.rename(options.collection, {
        dropTarget: true,
      });
      // update logs
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'aggregated';
      await connection.updateImportationLog(logs);
      // update progress
      progress.sources = sources;
      progress.dateEnd = Date.now();
      progress.state = 'aggregated';
      await connection.setProgress(progress);
      await debug.info('Aggregation Done');
    } else {
      await debug.info(`Aggregation already up to date`);
    }
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'inSilicoFragments',
        connection,
        stack: e.stack,
      });
    }
  }
}
