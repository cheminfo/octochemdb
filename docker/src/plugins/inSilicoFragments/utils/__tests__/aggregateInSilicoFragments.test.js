import { test, expect } from 'vitest';

import { OctoChemConnection } from '../../../../utils/OctoChemConnection.js';
import { aggregate } from '../../aggregates/aggregate';

test('In silico fragmentation', async () => {
  const connection = new OctoChemConnection();
  const activesOrNaturalsCollection =
    await connection.getCollection('activesOrNaturals');

  while (true) {
    if ((await activesOrNaturalsCollection.countDocuments()) === 63) {
      break;
    }
  }

  await aggregate(connection);
  const collection = await connection.getCollection('inSilicoFragments');
  const result = await collection.findOne({
    _id: {
      noStereoTautomerID:
        'fnc@r@JRUipPQFQQJQYKIQSPiLEUmSUUTu@A@@Zh}`NOta`ZfOacxX',
      ionMode: 'positive',
      ionSource: 'esi',
    },
  });

  expect(result).toMatchSnapshot();
  await connection.close();
}, 100000);
