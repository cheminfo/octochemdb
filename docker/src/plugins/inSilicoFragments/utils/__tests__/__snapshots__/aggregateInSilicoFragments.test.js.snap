// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`In silico fragmentation 1`] = `
{
  "_id": "fnc@r@JRUipPQFQQJQYKIQSPiLEUmSUUTu@A@@Zh}\`NOta\`ZfOacxX",
  "data": {
    "masses": {
      "positive": [
        341.07864239965096,
        311.0680777156209,
        233.02112752224093,
        136.99999815418093,
        215.04694834756094,
        197.03638366353093,
        323.0680777156209,
        305.05751303159093,
        227.01056283821092,
        245.02112752224093,
        177.05462056887094,
        99.00767037549093,
        221.03638366353093,
        195.02073359907092,
        195.06518525290093,
        117.01823505952093,
        263.0316922062709,
        160.99999815418093,
        134.98434808972095,
        239.04694834756094,
        213.03129828310094,
        279.0863146209609,
      ],
    },
    "ocl": {
      "idCode": "fnc@r@JRUipPQFQQJQYKIQSPiLEUmSUUTs@A@@@",
    },
  },
}
`;
